pglock/core.py

```diff
@@ -0,0 +1,394 @@
+            avoid importing Python's timedelta object.
+        side_effect (str, default=pglock.RETURN): Adjust the runtime behavior.
+            ``pglock.RETURN`` will return the acquisition status when using the context manager.
+            ``pglock.RAISE`` will raise a ``django.db.utils.OperationalError`` if the lock cannot
+            be acquired or a timeout happens. ``pglock.SKIP`` will skip decoratored code if the
+            lock cannot be acquired or a timeout happens.
+
+    Returns:
+        bool: When using the default side effect, returns ``True`` if the lock was acquired or
+        ``False`` if not.
+
+    Raises:
+        django.db.utils.OperationalError: When a lock cannot be acquired or a timeout happens
+            when using ``side_effect=pglock.RAISE``.
+        ValueError: If an invalid ``side_effect`` is provided or no lock ID is supplied for
+            the context manager.
+        TypeError: If the lock ID is not a string or int.
+    """  # noqa
+
+    def __init__(
+        self,
+        lock_id=None,
+        *,
+        shared=False,
+        wait=True,
+        using=DEFAULT_DB_ALIAS,
+        timeout=None,
+        side_effect=RETURN,
+    ):
+        """Acquire an advisory lock"""
+        if side_effect not in (RETURN, RAISE, SKIP):
+            raise ValueError(
+                "side_effect must be one of pglock.RETURN, pglock.RAISE, or pglock.SKIP"
+            )
+
+        self.acquire = f'pg{"_try" if not wait else ""}_advisory_lock{"_shared" if shared else ""}'
+        self.release = f'pg_advisory_unlock{"_shared" if shared else ""}'
+        self.lock_id = lock_id
+        self.using = using
+        self.timeout = timeout
+        self.side_effect = side_effect
+        self.wait = wait
+        self.shared = shared
+        self._used_as_context_manager = True
+
+    @property
+    def int_lock_id(self):
+        return _cast_lock_id(self.lock_id)
+
+    def __call__(self, func):
+        self._used_as_context_manager = False
+
+        if not self.lock_id:
+            module = inspect.getmodule(func)
+            self.lock_id = f"{module.__name__}.{func.__name__}"
+
+        @functools.wraps(func)
+        def inner(*args, **kwargs):
+            with self._recreate_cm():
+                if self.acquired or self.side_effect != SKIP:
+                    return func(*args, **kwargs)
+
+        return inner
+
+    def __enter__(self):
+        if not self.lock_id:
+            raise ValueError("Must supply a lock ID as the first argument.")
+
+        if self._used_as_context_manager and self.side_effect == SKIP:
+            raise ValueError(
+                "Cannot use pglock.SKIP in a context manager."
+                " Use the @pglock.advisory decorator instead."
+            )
+
+        sql = f"SELECT {self.acquire}({self.int_lock_id})"
+
+        with connections[self.using].cursor() as cursor:
+            try:
+                with contextlib.ExitStack() as stack:
```

Comment:
Yea exactly, it's just IMO a cleaner way of doing conditional context managers. It's there to avoid code like this:

```
if self.side_effect == RETURN ...:
    with lock_timeout, atomic():
        yield
else:
    yield
```